File: java/extentreports/src/main/java/com/relevantcodes/extentreports/converters/LogConverter.java
package com.relevantcodes.extentreports.converters;

import java.util.ArrayList;
import java.util.List;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import com.relevantcodes.extentreports.LogSettings;
import com.relevantcodes.extentreports.LogStatus;
import com.relevantcodes.extentreports.model.Log;
import com.relevantcodes.extentreports.utils.DateTimeUtil;

public class LogConverter extends LogSettings {
	private Element test;
	private Boolean isChildNode = false;
	
	public List<Log> getLogList() {
		 Elements logList = isChildNode 
				? test.select(".collapsible-body tbody > tr") 
				: test.select(".test-body > .test-steps > table > tbody > tr");
        
        ArrayList<Log> extentLogList = new ArrayList<Log>();
        Log extentLog;
        
        for (Element log : logList) {
            extentLog = new Log();
            
            extentLog.setTimestamp(
                    DateTimeUtil.getDate(
                            log.select(".timestamp").first().text(), 
                            getLogTimeFormat()
                    )
            );
            
            if (log.select(".step-name").size() == 1) {
                extentLog.setStepName(log.select(".step-name").first().text());
            }
            
            LogStatus status = getStatusStringMarkup(log.select(".status").first());
            extentLog.setLogStatus(status);
            
            extentLog.setDetails(log.select(".step-details").first().html());
            
            extentLogList.add(extentLog);
        }
        
        return extentLogList;
	}
	
	private LogStatus getStatusStringMarkup(Element log) {
		if (log.hasClass("pass")) {
			return LogStatus.PASS;
		}
		else if (log.hasClass("fail")) {
			return LogStatus.FAIL;
		}
		else if (log.hasClass("fatal")) {
			return LogStatus.FATAL;
		}
		else if (log.hasClass("error")) {
			return LogStatus.ERROR;
		}
		else if (log.hasClass("warning")) {
			return LogStatus.WARNING;
		}
		else if (log.hasClass("info")) {
			return LogStatus.INFO;
		}
		else if (log.hasClass("skip")) {
			return LogStatus.SKIP;
		}
		
		return LogStatus.UNKNOWN;
	}
	
	public LogConverter(Element test, Boolean isChildNode) {
		this.test = test;
		this.isChildNode = isChildNode;
	}
	
	public LogConverter(Element test) {
		this.test = test;
	}
}

File: dotnet/RelevantCodes.ExtentReports/RelevantCodes.ExtentReports/Converters/LogConverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using HtmlAgilityPack;

using RelevantCodes.ExtentReports.Model;

namespace RelevantCodes.ExtentReports.Converters
{
    class LogConverter : IConvertible<Log>
    {
        private HtmlNode _testStepsNode;

        public LogConverter(HtmlNode TestStepsNode)
        {
            _testStepsNode = TestStepsNode;
        }

        public List<Log> Convert()
        {
            List<Log> logList = null;

            var table = _testStepsNode.Descendants("table").Where(x => x.Descendants("tbody") != null & x.Descendants("tbody").Any());
            
            if (table != null & table.Any())
            {
                var tbody = table.First().Descendants("tbody").First();
                var tr = tbody.ChildNodes.Where(x => x.Name.Equals("tr"));

                logList = new List<Log>(); ;

                tr.ToList().ForEach(row => 
                {
                    var log = new Log();

                    var cellCount = row.ChildNodes.Where(x => x.Name.Equals("td")).Count();

                    var statusNode = row.SelectSingleNode(".//td[contains(@class, 'status')]");
                    var status = GetStatus(statusNode.GetAttributeValue("class", "unknown"));

                    var timestamp = row.SelectSingleNode(".//td[contains(@class, 'timestamp')]").InnerText;
                    var details = row.SelectSingleNode(".//td[contains(@class, 'step-details')]").InnerHtml;

                    if (cellCount == 4)
                    {
                        var stepName = row.SelectSingleNode(".//td[contains(@class, 'step-name')]").InnerHtml;
                        log.StepName = stepName;
                    }

                    log.Timestamp = System.Convert.ToDateTime(timestamp);
                    log.LogStatus = status;
                    log.Details = details;

                    logList.Add(log);
                });
            }

            return logList;
        }

        private LogStatus GetStatus(string HtmlClass)
        {
            HtmlClass = HtmlClass.ToLower();

            if (HtmlClass.Contains("pass")) return LogStatus.Pass;
            if (HtmlClass.Contains("fail")) return LogStatus.Fail;
            if (HtmlClass.Contains("fatal")) return LogStatus.Fatal;
            if (HtmlClass.Contains("error")) return LogStatus.Error;
            if (HtmlClass.Contains("warning")) return LogStatus.Warning;
            if (HtmlClass.Contains("skip")) return LogStatus.Skip;
            if (HtmlClass.Contains("info")) return LogStatus.Pass;

            return LogStatus.Unknown;
        }
    }
}
